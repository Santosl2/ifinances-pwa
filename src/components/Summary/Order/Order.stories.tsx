import { ComponentMeta, ComponentStory } from "@storybook/react";
import { Order } from "./Order";

export default {
  title: "Components/Summary/Order",
  component: Order,
  args: {
    amount: 500,
    title: "Entradas",
    icon: "income.svg",
    isLoading: false,
  },
} as ComponentMeta<typeof Order>;

const Template: ComponentStory<typeof Order> = (args) => <Order {...args} />;

export const Default = Template.bind({});
