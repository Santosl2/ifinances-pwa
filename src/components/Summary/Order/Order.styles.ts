import styled from "styled-components";

export const OrderWrapper = styled.div`
  padding: 1.2rem 1.5rem;
  border-radius: 0.2rem;
  background: var(--shape);
  color: var(--text-title);
  box-shadow: 0 0 5px 0 rgba(0, 0, 0, 0.2);

  transition: transform 0.2s;

  cursor: pointer;

  &:hover {
    transform: translateY(-1rem);
  }
  strong {
    display: block;
    margin-top: 1rem;
    font-size: 2rem;
    font-weight: 500;
    line-height: 3rem;
  }
`;

export const OrderHeader = styled.header`
  display: flex;

  img {
    margin-left: auto;
  }
`;
