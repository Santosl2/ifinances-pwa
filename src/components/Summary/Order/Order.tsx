/* eslint-disable @typescript-eslint/no-non-null-assertion */
/* eslint-disable unused-imports/no-unused-vars */
import { useEffect, useRef } from "react";

import { OrderHeader, OrderWrapper } from "./Order.styles";
import { OrderProps } from "./Order.types";

const OrderVariants = {
  visible: {
    opacity: 1,
    y: 0,
  },
  hidden: {
    opacity: 0,
    y: "50%",
  },
};

export function Order({ amount, icon, title }: OrderProps): JSX.Element {
  const moneyRef = useRef<HTMLSpanElement>(null);

  function animateNumber(speed = 100) {
    let timeout!: NodeJS.Timeout;

    const moneyDiv = moneyRef.current!;
    const actualValue = Number(moneyDiv.innerText?.replace("R$ ", ""));
    const endValue = Number(moneyRef.current!.dataset.amount);
    const time = endValue / speed;

    if (actualValue < endValue) {
      moneyDiv.innerText = Math.ceil(actualValue + time).toFixed(2);
      timeout = setTimeout(() => animateNumber(speed), 1);
      return;
    }

    moneyDiv.innerText = endValue
      .toFixed(2)
      .replace(/\d(?=(\d{3})+\.)/g, "$&,");

    if (timeout !== undefined) clearTimeout(timeout);
  }

  useEffect(() => {
    animateNumber(100);
  }, []);

  return (
    <OrderWrapper
      initial="hidden"
      animate="visible"
      variants={OrderVariants}
      transition={{ duration: 0 }}
      whileHover={{ y: "-16px" }}
    >
      <OrderHeader>
        <p>{title}</p>
        <img src={icon} alt="Imagem Order" />
      </OrderHeader>

      <strong>
        R${" "}
        <span ref={moneyRef} data-amount={amount}>
          0
        </span>
      </strong>
    </OrderWrapper>
  );
}
