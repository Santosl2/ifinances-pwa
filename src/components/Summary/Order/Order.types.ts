/* eslint-disable no-shadow */
export enum OrderIcons {
  total = "total.svg",
  income = "income.svg",
  outcome = "outcome.svg",
}

export interface OrderProps {
  title: string;
  amount: number;
  icon: OrderIcons;
  isLoading?: boolean;
}
