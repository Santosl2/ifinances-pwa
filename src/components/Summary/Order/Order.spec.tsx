import { render, screen, waitFor } from "@testing-library/react";

import { Order } from "./Order";
import { OrderIcons } from "./Order.types";

describe("Order Test", () => {
  it("should render Order", async () => {
    render(<Order amount={1000} icon={OrderIcons.income} title="Teste" />);

    expect(screen.getByText("Teste")).toBeInTheDocument();
    expect(screen.getByAltText("Imagem Order")).toBeInTheDocument();
  });

  it("should be able to render Skeleton", async () => {
    render(
      <Order amount={1000} icon={OrderIcons.income} title="Teste" isLoading />
    );

    expect(screen.getByTestId("skeleton-test")).toBeInTheDocument();
  });

  it("should be match snapshot", () => {
    const { container } = render(
      <Order amount={1000} icon={OrderIcons.income} title="Teste" />
    );

    expect(container.firstChild).toMatchSnapshot();
  });
});
