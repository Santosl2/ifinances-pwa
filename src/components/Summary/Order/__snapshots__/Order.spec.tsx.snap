// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Order Test should be match snapshot 1`] = `
<div
  class="sc-gsnTZi jsbwHh"
  style="opacity: 0; transform: translateY(50%) translateZ(0);"
>
  <header
    class="sc-dkzDqf hpuwwA"
  >
    <p>
      Teste
    </p>
    <img
      alt="Imagem Order"
      src="income.svg"
    />
  </header>
  <strong>
    R$
    <span
      data-amount="1000"
    >
      0
    </span>
  </strong>
</div>
`;
